dpll_mon.c

```diff
@@ -0,0 +1,991 @@
+	if (tb[DPLL_A_MODE])
+		dm->dpll_mode = nla_get_u32(tb[DPLL_A_MODE]);
+	if (tb[DPLL_A_LOCK_STATUS])
+		dm->lock_status = nla_get_u32(tb[DPLL_A_LOCK_STATUS]);
+}
+
+static int dpll_mon_device_owner(struct dpll_mon *dm, struct nlattr **tb)
+{
+	if (tb[DPLL_A_CLOCK_ID] && tb[DPLL_A_MODULE_NAME] && tb[DPLL_A_TYPE])
+		if (nla_get_u64(tb[DPLL_A_CLOCK_ID]) == dm->clock_id &&
+		    !nla_strcmp(tb[DPLL_A_MODULE_NAME],
+				dm->module_name) &&
+		    nla_get_u32(tb[DPLL_A_TYPE]) == DPLL_TYPE_EEC)
+			return 1;
+
+	return 0;
+}
+
+static int pins_ready(struct dpll_mon *dm)
+{
+	struct dpll_mon_pin *pin;
+
+	STAILQ_FOREACH(pin, &dm->pins, list) {
+		if (!pin->ready) {
+			dm->init_err++;
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+static int pins_dnu(struct dpll_mon *dm)
+{
+	struct dpll_mon_pin *pin;
+
+	STAILQ_FOREACH(pin, &dm->pins, list) {
+		if (pin->prio_valid && pin->prio != dm->dev_dnu_prio) {
+			pr_debug("not expected pin id:%u prio: %u",
+				 pin->id, pin->prio);
+			dm->init_err++;
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+static void invalidate_pins(struct dpll_mon *dm)
+{
+	struct dpll_mon_pin *pin;
+
+	STAILQ_FOREACH(pin, &dm->pins, list)
+		pin->ready = 0;
+}
+
+static int dpll_mon_ntf_recv(struct nl_msg *msg, void *arg)
+{
+	struct nlattr *tb[MAX_ATTR + 1];
+	struct dpll_mon *dm = arg;
+	struct genlmsghdr *gnlh;
+	int ret;
+
+	gnlh = nlmsg_data(nlmsg_hdr(msg));
+	if (!gnlh) {
+		pr_err("failed parse gnl header");
+		return -EINVAL;
+	}
+	ret = genlmsg_parse(nlmsg_hdr(msg), 0, tb, MAX_ATTR, NULL);
+	if (ret) {
+		pr_err("genlsmg_parse err:%d", ret);
+		return ret;
+	}
+	switch (gnlh->cmd) {
+	case DPLL_CMD_DEVICE_CREATE_NTF:
+		if ((dm->state == DPLL_MON_STATE_INIT_READY ||
+		     dm->state == DPLL_MON_STATE_DEV_GET_WAIT) &&
+		    dpll_mon_device_owner(dm, tb)) {
+			dpll_mon_device_update(dm, tb);
+			dpll_mon_state_set(dm, DPLL_MON_STATE_DEV_INITED);
+		}
+		break;
+	case DPLL_CMD_DEVICE_CHANGE_NTF:
+		if (dpll_mon_device_owner(dm, tb))
+			dpll_mon_device_update(dm, tb);
+		break;
+	case DPLL_CMD_DEVICE_DELETE_NTF:
+		if (dpll_mon_device_owner(dm, tb))
+			dpll_mon_state_set(dm, DPLL_MON_STATE_DPLL_REINIT);
+		break;
+	case DPLL_CMD_PIN_DELETE_NTF:
+		dpll_mon_pin_update(dm, tb, 0, 1, gnlh);
+		break;
+	case DPLL_CMD_PIN_CREATE_NTF:
+	case DPLL_CMD_PIN_CHANGE_NTF:
+		dpll_mon_pin_update(dm, tb, 1, 1, gnlh);
+		break;
+	default:
+		break;
+	}
+
+	return NL_OK;
+}
+
+static int dpll_rt_recv(struct nl_msg *msg, void *arg)
+{
+	struct nlattr *tb[RTM_MAX + 1], *an;
+	struct dpll_mon *dm = arg;
+	struct dpll_mon_pin *pin;
+	struct ifinfomsg *info;
+	struct nlmsghdr *nlh;
+	int ret, rem;
+
+	nlh = nlmsg_hdr(msg);
+	if (!nlh) {
+		pr_err("failed parse nl header");
+		return -EINVAL;
+	}
+	ret = nlmsg_parse(nlmsg_data(nlh), sizeof(*info), tb, IFLA_MAX, NULL);
+	if (ret) {
+		pr_err("rt nlsmg_parse err:%d", ret);
+		return ret;
+	}
+	info = nlmsg_data(nlh);
+	pin = find_pin_by_if_index(dm, info->ifi_index);
+	if (tb[IFLA_DPLL_PIN]) {
```

Comment:
Just nitpicking, but we might consider flattening it a bit for readability, e.g. lines 532 to 544 might be:
```
	if (!tb[IFLA_DPLL_PIN])
		return NL_OK;

	nla_for_each_nested(an, tb[IFLA_DPLL_PIN], rem) {
		if (nla_type(an) != DPLL_A_PIN_ID)
			continue;
		if (pin) {
			pin->id = nla_get_u32(an);
			continue;
		}
		pin = find_pin(dm, nla_get_u32(an));
		if (pin)
			pin->if_index = info->ifi_index;
	}
``` 
From 4 nesting levels we have only 2, but I don't really insist on changing that. :)

Reply:
this was already changed a bit, no longer need to read for each nested, as there is only one nest per netdevice, but yeah will double check if this can be flattened as suggested, thanks!